### ISSUE TITLE
Add radial fade

### ISSUE THREAD

**Issue by mrjustaguy** [godotengine/godot-proposals#2963]
### Describe the project you are working on
Stealth Game

### Describe the problem or limitation you are having in your project
Want to Fade things a certain radius away from the camera, while keeping the sky

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
Add an Environment/Camera/Material feature that works like Pixel Alpha Distance Fade, just instead of making a Linear cut, making it Radial

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
What is desired is the Red Fog part to not be rendered (be transparent)
![Fade](https://user-images.githubusercontent.com/50302479/124928510-f0878580-dfff-11eb-9337-f65fc08e2d88.png)

Note: This is a very sharp cut off in this example, but it should also have a curve, much like the depth curve in the fog, so that the user can decide if they want a slow fade away, or a very sudden fade away, so like just copying the Fog's Depth settings, but instead of applying color, making things transparent...
This could be added as an option to the existing Fog in Godot 3, but would possibly be incompatible with Godot 4's Fog as there the fog uses density, which doesn't allow for the level of fine control that Godot 3 offers with fogs (but is more realistic otherwise)

### If this enhancement will not be used often, can it be worked around with a few lines of script?
Only one I can think of is making a Sphere that's rendering the sky, but that has it's own issues.

### Is there a reason why this should be core and not an add-on in the asset library?
Rendering is core.

**Comment by Calinou** [comment]
This is already handled by the **Aerial Perspective** environment fogh property in the `master` branch. I doubt it can be backported easily to `3.x` though.

**Comment by clayjohn** [comment]
@Calinou Yeah, it shouldn't be too hard, it is a pretty simple hack. Instead of using a fixed colour for the background fog, you look up the colour from the radiance cubemap (the sky cubemap that is used for reflections). 

I'm not sure aerial perspective is what the OP is asking for though, it sounds like they just want to have distance fade that curves more at the edges of the screen. In that case the solution is much simpler as distance fade is a single line of code. You just need to replace ``-VERTEX.z`` with something a little more curved. 

```
ALPHA *= clamp(smoothstep(distance_fade_min, distance_fade_max, -VERTEX.z), 0.0, 1.0);
```

**Comment by Calinou** [comment]
> I'm not sure aerial perspective is what the OP is asking for though, it sounds like they just want to have distance fade that curves more at the edges of the screen. In that case the solution is much simpler as distance fade is a single line of code. You just need to replace -VERTEX.z with something a little more curved.

Sounds good to me, but having a more curved falloff may have a performance cost.

**Comment by mrjustaguy** [comment]
The Circular radius could be a toggle

**Comment by Calinou** [comment]
I got aerial perspective fog working in `3.x` GLES3:

https://user-images.githubusercontent.com/180032/124987733-7b3aa580-e03d-11eb-93c1-b718880fe7c6.mp4

The way it's implemented isn't very clean (and is hardcoded), so I won't open a pull request yet.

**Comment by atirut-w** [comment]
Any news regarding this? It's been so long since the last comment.

**Comment by Calinou** [comment]
> Any news regarding this? It's been so long since the last comment.

I don't have time available to continue working on this, sorry.

**Comment by elvisish** [comment]
I don't think this is a duplicate of this :https://github.com/godotengine/godot-proposals/issues/7138  as this is requesting using canvas as fog albedo, not a panorama/cubemap? The two perspectives would look significantly different (canvas is a 2D projection and a panorama would be obviously spherical).

**Comment by elvisish** [comment]
> I got aerial perspective fog working in `3.x` GLES3:
> 
> 
> 
> https://user-images.githubusercontent.com/180032/124987733-7b3aa580-e03d-11eb-93c1-b718880fe7c6.mp4
> 
> 
> 
> The way it's implemented isn't very clean (and is hardcoded), so I won't open a pull request yet. If anyone wants to continue this work, the branch is here: https://github.com/Calinou/godot/tree/environment-fog-add-aerial-perspective
> 
> 
> 
> It may be implementable in GLES2 too, although I haven't investigated that yet.
> 
> 
> 
> **Testing project (made for the branch linked above):** [test_aerial_perspective_3.x.zip](https://github.com/godotengine/godot-proposals/files/6787146/test_aerial_perspective_3.x.zip)

Is it possible to reimplement fog as a script or extension so canvas could be used as the fog color without having to port it to worldenvironment?

**Comment by Calinou** [comment]
> Is it possible to reimplement fog as a script, shader or extension so canvas could be used as the fog color without having to port it to worldenvironment?

It's possible to draw fog using a material shader, but you need to apply this custom shader on every 3D mesh.

**Comment by elvisish** [comment]
> > Is it possible to reimplement fog as a script, shader or extension so canvas could be used as the fog color without having to port it to worldenvironment?
> 
> 
> 
> It's possible to draw fog using a material shader, but you need to apply this custom shader on every 3D mesh.
> 
> 
> 
> Post-processing-based fog using the depth texture is also an option, but I believe it has different issues (such as not handling transparent objects correctly).

I'm only using alpha clip for transparency, would this just work as a screen shader (on a colorRect) or would it also need to be on each mesh?
